### PR TITLE
Remove shape_inference from initialization of ONNXGraph

### PR DESCRIPTION
**Pull request by kshpv**
### Changes

Remove shape_inference call from initialization of **ONNXGraph**. 
Also, remove unnecessary dependency for ONNX -  **skl2onnx**

### Reason for changes

**ONNXGraph** aims to provide the interface for obtaining necessary information from the ONNX model. 

One of the functionality is to provide the information about activation tensors. Typically, the ONNX model doesn't contain any info about activatiion tensors. To get it this is essential to do shape inference. 
Currently, the shape inference is done for every ONNX model during initialization. This PR implements a lazy initialization for ONNXGraph, meaning to do shape_inference only when it is needed. 

This speeds up the quantization pipeline much.

### Related tickets

No

### Tests

No

### CODE REVIEW

File: nncf/experimental/onnx/graph/onnx_graph.py [hunk 1]
```diff
@@ -33,38 +32,21 @@ class ONNXGraph:
 
     def __init__(self, onnx_model: onnx.ModelProto):
```
Comment (andrey-churkin) [inline nncf/experimental/onnx/graph/onnx_graph.py:33]: Is it possible to make all methods inside the `ONNXGraph` static (i.e. add `@staticmethod`)?

Comment (vshampor) [reply nncf/experimental/onnx/graph/onnx_graph.py:33]: We've just discussed this with @kshpv. Having all methods in a class as static really makes you consider whether you need a class at all. I suppose that here we do need a class, namely - its objects as entities with independent states and lifetimes.

Comment (andrey-churkin) [reply nncf/experimental/onnx/graph/onnx_graph.py:33]: @kshpv Is there at least one method inside the `ONNXGraph` that modifies the state of the object (i.e. `self`)?

As far as I can see this class looks like a collection of independent methods (utilities) for working with `onnx.ModelProto`.

Comment (kshpv) [reply nncf/experimental/onnx/graph/onnx_graph.py:33]: > @kshpv Is there at least one method inside the `ONNXGraph` that modifies the state of the object (i.e. `self`)?
> 
> As far as I can see this class looks like a collection of independent methods (utilities) for working with `onnx.ModelProto`.

ONNXGraph provides the methods to get the necessary infromation from the ONNX model. It doesn't change it. 
To apply transformation on the ONNX model, we have ONNXModelTransformer.

Comment (andrey-churkin) [reply nncf/experimental/onnx/graph/onnx_graph.py:33]: So, I still have my question. Why we can't decorate all methods inside the ONNXGraph as `@staticmethod`? Or better, remove this class and create the module with all methods that regard the `onnx.ModelProto`. For example, `nncf/experimental/onnx/graph/utils.py`

Comment (kshpv) [reply nncf/experimental/onnx/graph/onnx_graph.py:33]: Your approach is possible, I implemented it as the first iteration. 
The problem here is that the developer should control the correct statement of the ONNX model during quantization, meaning to do infer_shape while it is necessary. 
I am working on another one, where ONNXGraph stores the model and does infer shape only when it is necessary. In that case, the developer doesn't have to care about the correct statement of ONNX model.

Comment (kshpv) [reply nncf/experimental/onnx/graph/onnx_graph.py:30]: I've pushed the new version. @andrey-churkin, please, take a look

Comment (kshpv) [reply nncf/experimental/onnx/graph/onnx_graph.py:30]: @vshampor, please also take a look

**Comment by kshpv** [comment]
I've updated the description of the PR

File: nncf/experimental/onnx/graph/onnx_graph.py [hunk 2]
```diff
@@ -79,45 +78,51 @@ def get_model_inputs(self) -> List[ValueInfoProto]:
...
-                        'output': list(node.output)}
+        if self._node_name_to_node is None:
+            self._update_node_names()
+        if node_name in self._node_name_to_node:
+            return {'input': list(self._node_name_to_node[node_name].input),
+                    'output': list(self._node_name_to_node[node_name].output)}
         raise RuntimeError('There is no node with the name {}'.format(node_name))
 
-    def get_input_port_id_for_node_after_input(self, input_name: str, to_node: NodeProto) -> int:
+    def get_input_port_id_for_node_after_input(self, input_name: str, to_node: onnx.NodeProto) -> int:
```
Comment (KodiaqQ) [inline nncf/experimental/onnx/graph/onnx_graph.py:125]: Why don't we make this method static?

File: nncf/experimental/onnx/graph/onnx_graph.py [hunk 3]
```diff
@@ -129,7 +134,7 @@ def get_input_port_id_for_node_after_input(self, input_name: str, to_node: NodeP
                 return input_port_id
         raise RuntimeError(f'The node {to_node} does not have input edge with the name {input_name}')
 
-    def get_output_port_id_for_node_before_output(self, output_name: str, from_node: NodeProto) -> int:
+    def get_output_port_id_for_node_before_output(self, output_name: str, from_node: onnx.NodeProto) -> int:
```
Comment (KodiaqQ) [inline nncf/experimental/onnx/graph/onnx_graph.py:137]: Same.

File: nncf/experimental/onnx/graph/onnx_graph.py (hunk 2 above)
Comment (kshpv) [reply nncf/experimental/onnx/graph/onnx_graph.py:125]: yes, we could

File: nncf/experimental/onnx/graph/onnx_graph.py (hunk 3 above)
Comment (kshpv) [reply nncf/experimental/onnx/graph/onnx_graph.py:137]: I'll update it

File: nncf/experimental/onnx/graph/onnx_graph.py [hunk 4]
```diff
@@ -248,24 +249,44 @@ def get_tensor_shape(self, tensor: ValueInfoProto) -> List[int]:
...
+        If after updating of the self._activations_tensor_name_to_value_info, there is still no such tensor,
+        do shape inference of the model.
+        :param edge_name: The name of the edge.
+        :return: Shape of the tensor on that edge.
+        """
+        if self._activations_tensor_name_to_value_info is None:
+            self._update_activation_tensors()
+        if edge_name in self._activations_tensor_name_to_value_info:
+            return self.get_tensor_shape(self._activations_tensor_name_to_value_info[edge_name])
+        self._update_activation_tensors(do_shape_inference=True)
```
Comment (vshampor) [inline nncf/experimental/onnx/graph/onnx_graph.py:263]: Is this right? I don't think that doing shape inference at every call of `get_edge_shape` is efficient, or lazy.

File: nncf/experimental/onnx/graph/onnx_graph.py [hunk 5]
```diff
@@ -248,24 +249,44 @@ def get_tensor_shape(self, tensor: ValueInfoProto) -> List[int]:
...
+        If after updating of the self._activations_tensor_name_to_value_info, there is still no such tensor,
+        do shape inference of the model.
+        :param edge_name: The name of the edge.
+        :return: Shape of the tensor on that edge.
+        """
+        if self._activations_tensor_name_to_value_info is None:
+            self._update_activation_tensors()
+        if edge_name in self._activations_tensor_name_to_value_info:
+            return self._activations_tensor_name_to_value_info[edge_name].type.tensor_type.elem_type
+        self._update_activation_tensors(do_shape_inference=True)
```
Comment (vshampor) [inline nncf/experimental/onnx/graph/onnx_graph.py:281]: Same

**Review by vshampor — CHANGES_REQUESTED** [review]

File: nncf/experimental/onnx/graph/onnx_graph.py (hunk 4 above)
Comment (vshampor) [reply nncf/experimental/onnx/graph/onnx_graph.py:263]: Sorrz, missed the early exit clause.

**Review by alexsu52 — APPROVED** [review]
LGTM

File: nncf/experimental/onnx/graph/onnx_graph.py [hunk 6]
```diff
@@ -11,64 +11,63 @@
...
-        try:
-            return self.lookup_nodes[node_name]
-        except KeyError as e:
-            raise KeyError(f"There is no node with the name {node_name}") from e
+    def get_node_by_name(self, node_name: str) -> Optional[onnx.NodeProto]:
+        """
+        Returns a model node with the name equals to 'node_name' from self._node_name_to_node.
+        If the self._node_name_to_node is None, fills it with the nodes from the self.onnx_model.
+        If there is no node with such name returns None.
+        :param node_name: Name of the node.
```
[the remark this answers is not in the document]
Comment (kshpv) [reply nncf/experimental/onnx/graph/onnx_graph.py]: Aligned with styleguide

File: nncf/experimental/onnx/graph/onnx_graph.py (hunk 3 above)
Comment (kshpv) [reply nncf/experimental/onnx/graph/onnx_graph.py]: Done

File: nncf/experimental/onnx/graph/onnx_graph.py (hunk 2 above)
Comment (kshpv) [reply nncf/experimental/onnx/graph/onnx_graph.py]: Done

**Review by vshampor — APPROVED** [review]

**Review by KodiaqQ — APPROVED** [review]

File: nncf/experimental/onnx/graph/onnx_graph.py (hunk 1 above)
Comment (AlexKoff88) [reply nncf/experimental/onnx/graph/onnx_graph.py:30]: I've looked as well and suggest merging this PR.

**Review by AlexKoff88 — APPROVED** [review]

**Review by andrey-churkin — APPROVED** [review]

**Pull request merged**